contracts/CommunityVault.sol

```diff
@@ -14,15 +16,91 @@ import { CommunityOwnable } from "./CommunityOwnable.sol";
  *      Only community owners, as defined in the CommunityOwnable contract, have
  *      permissions to transfer these tokens out of the vault.
  */
-contract CommunityVault is CommunityOwnable {
+contract CommunityVault is CommunityOwnable, IERC721Receiver {
     using SafeERC20 for IERC20;
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    event ERC20Deposited(address indexed depositor, address indexed token, uint256 amount);
+    event ERC721Deposited(address indexed depositor, address indexed token, uint256 tokenId);
 
     error CommunityVault_LengthMismatch();
     error CommunityVault_NoRecipients();
     error CommunityVault_TransferAmountZero();
+    error CommunityVault_ERC20TransferAmountTooBig();
+    error CommunityVault_DepositAmountZero();
+    error CommunityVault_IndexOutOfBounds();
+    error CommunityVault_ERC721TokenAlreadyDeposited();
+    error CommunityVault_ERC721TokenNotDeposited();
+
+    mapping(address => uint256) public erc20TokenBalances;
+    mapping(address => EnumerableSet.UintSet) private erc721TokenIds;
 
     constructor(address _ownerToken, address _masterToken) CommunityOwnable(_ownerToken, _masterToken) { }
 
+    /**
+     * @dev Allows anyone to deposit ERC20 tokens into the vault.
+     * @param token The address of the ERC20 token to deposit.
+     * @param amount The amount of tokens to deposit.
+     */
+    function depositERC20(address token, uint256 amount) external {
+        if (amount == 0) {
+            revert CommunityVault_DepositAmountZero();
+        }
+
+        // Transfer tokens from the sender to this contract
+        IERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+
+        // Update the total balance of the token in the vault
+        erc20TokenBalances[token] += amount;
+
+        // Emit an event for the deposit (optional, but recommended for tracking)
+        emit ERC20Deposited(msg.sender, token, amount);
+    }
+
+    /**
+     * @dev Allows anyone to deposit multiple ERC721 tokens into the vault.
+     * @param token The address of the ERC721 token to deposit.
+     * @param tokenIds The IDs of the tokens to deposit.
+     */
+    function depositERC721(address token, uint256[] memory tokenIds) public {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            // Transfer the token from the sender to this contract
+            IERC721(token).safeTransferFrom(msg.sender, address(this), tokenIds[i]);
+
+            // Add the token ID to the EnumerableSet for the given token
+            bool added = erc721TokenIds[token].add(tokenIds[i]);
+            if (!added) {
+                revert CommunityVault_ERC721TokenAlreadyDeposited();
+            }
```

Comment:
Shouldn't we perform this check before we even try to transfer it?

Reply:
good idea, I moved it